research/automated_release.md

```diff
@@ -0,0 +1,148 @@
+
+`master` is the branch intended for stable software releases. This branch should only receive code that is thoroughly tested, and does not contain partially finished features.
+
+In order to verify that code on `develop` is ready for primetime, new `release` branches are created regularly. These fork from `develop` either after a set interval (end of Scrum sprint), or when a feature is complete.
+
+At this point developers can continue to add new features to `develop`, while the `release` branch is used to verify that its frozen set of features are bug-free, and meet requirements.
+
+Bug fixes made on `release` branches should also be merged into develop. No new functionality should be added to `release`.
+
+When the software on `release` passes all checks, it is merged into `master`.
+
+Occasionally, despite best efforts, `master` will have bugs. For the most critical of them, `hotfix` branches are forked directly from `master` to fix the issue.
+When ready, these branches are forked both into `master`, and `develop`.
+
+Brewblox as a project has no special requirements that mandate any changes to this model. At time of writing, we do still need to implement meaningful and automatic regression tests that should be run against `release` software.
+
+We do need to formulate a standard approach for conditions not described explicitly by Git Flow:
+
+* Contributors working on forked repositories
+* Pull Requests
+
+## Git Flow: Forked
+
+The default Git Flow model does not describe forked repositories making upstream pull requests. The forked model is attractive to projects with open-source contributors, as it allows contributions without granting write access to strangers.
+
+Contributors fork the base repository on GitHub into their own. At that point they can create their own branches, or use their local version of `develop` - the upstream repository doesn't care. When they are done, they make a pull request to the correct target branch on the base repository. Pull requests can be commented on by many, but only merged by somebody with write access to the base repository.
+
+We'll refer to the base repository as `upstream`, and to the contributor repository as `downstream`.
+
+In a forked Git Flow repository, there will be no `feature` or `hotfix` branches in `upstream`. This is all done in `downstream`.
+
+Contributors fork new `downstream/feature` branches from `upstream/develop`, and when ready, make a pull request back into `upstream/develop`.
+
+`release` branches are created in `upstream`. If any fixes need to be made, these are directly done in `downstream/release` (or a forked branch), and merged back in through pull request.
+
+The same goes for `hotfix` branches: `downstream/hotfix` is forked from `upstream/master`, and merged back into `upstream/master`.
+
+At any given time, there are at most three active branches in `upstream`: `master`, `develop`, and (when active) `release`.
```

Comment:
In general, only 3 branches will exist. Exceptions are:
- Long running feature branches that involve multiple developers. They can use a feature branch on the upstream repo to synchronize work. They could do PRs to the feature branch.
- Multiple release branches can exist in parallel, for example the latest stable release and a new release candidate for the next major version.

In general: small features do not require a feature branches on the upstream repo.

Reply:
Fixed